pkg/agent/factsopenshift.go

```diff
@@ -0,0 +1,110 @@
+package agent
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"time"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+)
+
+type OpenshiftVersionDesired struct {
+	Version string
+}
+
+type OpenshiftVersionHistory struct {
+	State          string
+	Verified       bool
+	Version        string
+	CompletionTime time.Time
+}
+
+type OpenshiftVersionStatus struct {
+	Desired OpenshiftVersionDesired
+	History []OpenshiftVersionHistory
+}
+type OpenshiftVersion struct {
+	Status OpenshiftVersionStatus
+}
+
+func (col factCollector) fetchOpenshiftVersion(ctx context.Context) (*OpenshiftVersionFact, error) {
+	body, err := col.client.RESTClient().Get().AbsPath("/apis/config.openshift.io/v1/clusterversions/version").Do(ctx).Raw()
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// API server doesn't know `clusterversions` or there is no resource, so we are not running on openshift.
+			return nil, nil
+		}
+		return nil, err
+	}
+	var version OpenshiftVersion
+	err = json.Unmarshal(body, &version)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse the openshift version: %w", err)
+	}
+
+	return processOpenshiftVersion(version)
+}
+
+type SemanticVersion struct {
+	Major string
+	Minor string
+	Patch string
+}
+
+type OpenshiftVersionFact struct {
+	Version        SemanticVersion
+	DesiredVersion SemanticVersion
+}
+
+func processOpenshiftVersion(v OpenshiftVersion) (*OpenshiftVersionFact, error) {
+	currentVersion := ""
+	lastedUpdate := time.Time{}
+	for _, h := range v.Status.History {
+		if h.State == "Completed" && h.Verified == true && h.CompletionTime.After(lastedUpdate) {
+			currentVersion = h.Version
+			lastedUpdate = h.CompletionTime
+		}
+	}
+	var err error
+	versionFact := &OpenshiftVersionFact{}
+	versionFact.DesiredVersion, err = parseSematicVersion(v.Status.Desired.Version)
+	if err != nil {
+		return versionFact, fmt.Errorf("unable to parse desiredVersion: %w", err)
+	}
+	versionFact.Version, err = parseSematicVersion(currentVersion)
+	if err != nil {
+		return versionFact, fmt.Errorf("unable to parse version: %w", err)
+	}
```

Comment:
From what I can tell, we currently don't store the OCP version fact if we cannot parse the current version. I don't remember if we discussed this and decided that we only want to report the OCP version if we can parse both the desired and current version.

Also, does the current version parsing work correctly on a freshly installed cluster where we've never done an update?

Reply:
You're right, not storing anything if we cannot parse one of the versions is probably not the right approach. I'll fix this.

As for the fresh cluster: I think that actually might be empty on a fresh cluster.. https://docs.openshift.com/container-platform/4.9/rest_api/config_apis/clusterversion-config-openshift-io-v1.html

I guess in that case we could fall back to the desired version? Or do you have a better idea?